homeassistant/components/switch/mochad.py

```diff
@@ -59,17 +64,37 @@ def name(self):
 
     def turn_on(self, **kwargs):
         """Turn the switch on."""
-        self._state = True
+        from pymochad.exceptions import MochadException
+        _LOGGER.debug("Reconnect %s:%s", self._controller.server,
+                      self._controller.port)
         with mochad.REQ_LOCK:
-            self.device.send_cmd('on')
-            self._controller.read_data()
+            try:
+                # Recycle socket on new command to recover mochad connection
+                self._controller.reconnect()
+                self.device.send_cmd('on')
+                # No read data on CM19A which is rf only
+                if self._comm_type == 'pl':
+                    self._controller.read_data()
+                self._state = True
+            except (MochadException, OSError) as exc:
+                _LOGGER.error("Error with mochad communication: %s", exc)
 
     def turn_off(self, **kwargs):
         """Turn the switch off."""
-        self._state = False
+        from pymochad.exceptions import MochadException
+        _LOGGER.debug("Reconnect %s:%s", self._controller.server,
+                      self._controller.port)
         with mochad.REQ_LOCK:
-            self.device.send_cmd('off')
-            self._controller.read_data()
+            try:
+                # Recycle socket on new command to recover mochad connection
+                self._controller.reconnect()
+                self.device.send_cmd('off')
+                # No read data on CM19A which is rf only
+                if self._comm_type == 'pl':
+                    self._controller.read_data()
+                self._state = False
+            except (MochadException, OSError) as exc:
+                _LOGGER.error("Error with mochad communication: %s", exc)
 
     def _get_device_status(self):
```

Comment:
It seems `_get_device_status` can be used to update state, but why isn't that called in an `update` method? Then the new state would automatically be polled after each service call.

Reply:
get_device_status is shortcut to the parent SwitchDevice.get_status(). It doesn't make a trip to Mochad and the device.  Rather it's a status of the device in hass. No need to change anything here.  Not my code. But I did a little research. Mochad supports status_request for pl commands, but not for rf. https://bfocht.github.io/mochad/mochad_reference.html  I think the reasoning behind this original code is as I described in my above comment.

Reply:
It does ask the mochad device for status.

Reply:
Yes, you're correct. I might invest more time into all this x10 stuff at some later point. For now, whatever changes I made here, work for my home with ca 15 x10 devices of different kinds. I can also do bug fix if anyone reports an issue.